API.md

```diff
@@ -0,0 +1,229 @@
+  var key = new ConnectionKey(packet.GetSourceNode(), packet.GetDestinationNode());
+  NatConnection<TPacket,TNode> connection;
+  if (NAT_MapToInside.TryGetValue(key, out connection))
+  {
+    var destination = connection.InsideNode;
+
+    // Update any connection state, including resetting the inactivity timer
+    connection.ReceivedPacket(packet, packetFromInside: false);
+
+    // Rewrite the packet destination
+    packet.SetDestination(destination);
+
+    // Update checksums
+    packet.UpdateChecksums();
+
+    // Forward on the mapped network port
+    return new ForwardingDecision.SinglePortForward(destination.InterfaceNumber);
+  }
+  else
+  {
+    return Drop;
+  }
+}
+```
+
+
+## Details
+The API consists of 3 parts:
+* The [core API](#Interfaces)
+* Pax's [frontend](#Frontend)
+* [Version management](#Versioning)
+
+(FIXME: this doesn't yet cover the "Lite" API and PaxConfig.)
+
+
+### Frontend
+The *frontend* consists of the command-line tool that is passed a DLL and
+configuration, and that runs the packet processors in your OS.
+
+The code for this is in [Pax.cs](Pax.cs), in the class called `Frontend`.
+You can interact with this in two ways:
+* providing switches to the tool, to control its appearance perhaps, as if it was called from the command line.
+* terminating the Frontend, for batch processes for example, by calling [Frontend.shutdown](https://github.com/niksu/pax/blob/6cffc3edd2741896a068d7832a2b1d39a29589af/Pax.cs#L395).
+
+
+### FIXME: Versioning
+(Talk about static and dynamic checking)
+
+
+### Interfaces
+Pax provides different interfaces to facilitate the writing of different kinds
+of packet processors. The code for this can be found in
+[Paxifax.cs](Paxifax.cs).
+Remember that ultimately the programming model consists of reading/writing to
+logical network ports that are linked to resources made available by the OS
+(e.g., physical NICs). But it's useful to differentiate between whether a packet
+processor is allow to forward packets for example, and how many ports its
+allowed to forward to.
+
+#### Basic packet processors
+All packet processors in Pax implement at least one of two interfaces. These
+interfaces reflect the abstraction that there are two kinds of basic packet
+processors:
+* `IAbstract_PacketProcessor` implements a `process_packet` function that indicates on which logical interfaces to map a (possibly modified) packet to.
+* `IHostbased_PacketProcessor` implements a `packetHandler` functions that works through side-effects to forward packets, and that is not directly given a packet but rather needs to extract it from another argument.
+
+##### IAbstract_PacketProcessor
+Ideally packet processors should implement this interface, since it was intended
+to be used by packet processors that don't necessarily run on Pax's runtime
+(i.e., some other runtime environment).
+([Code](https://github.com/niksu/pax/blob/6cffc3edd2741896a068d7832a2b1d39a29589af/Paxifax.cs#L20))
```

Comment:
Yes partly to target a specific line number, and also to target a specific revision of the code (since in the most recent revision of the code the line number of a definition might have changed, which might be confusing for anybody who clicks on the link).

The approach I'm using is a bit brittle since API.md needs to be edited each time the API is changed, and it would be useful to have more automation. Perhaps a docsygen-like approach might be a good idea, but I'd like to retain more flexibility in sequencing/presenting the API description, rather than use the sequence in which definitions appear in the code. I prepared a patch which clarifies that the link points to a specific revision of the code (which might not be the current revision), as a cue to the reader in case the docs and code diverged.